# Group 6 SC1015 Assignment

We will be looking into Car Features and how it will affect the MSRP of the cars. <br>

This dataset was from kaggle [Car Features and MSRP](https://www.kaggle.com/CooperUnion/cardataset) <br>

Since a description of the data was not given. Let us list out the features and their details <br>
  - __Car Make__
      * Brand of the vehicle
  - __Car Model__
      * Can also be "Name" of car
  - __Car Year__
      * When the car was manufactured
  - __Engine Fuel Type__
      * Fuel type reccomendation for the car. If it is an electric vehicle "electric" will be shown.
  - __Engine Hourse Power(HP)__
      * Refers to the power an engine produces
  - __Engine Cylinders__
      * Cylinders are located in an engine, it is  a chamber where fuel is combusted and power is generated. In this data, it refers to the amount of cylinders the Engine has.
  - __Transimission Type__
      * The way power is transmitted from an engine to the axle in the vehicle
  - __Driven Wheels__
      * Different Vehicles run on different wheel drives. Front wheel drive cars will have power transmitted to the Front 2 wheels. Rear wheel drive cars will have power transmitted to the Rear 2 wheels. All wheel drive cars will have power transmitted to all the Wheels all the time. 4WD is similar to All wheel drive cars, just with the fact that it can toggle between 4WD and 2WD. 
  - __Number of Doors__
      * Self explanatory
  - __Market Category__
      * Refers to the market audience for the cars
  - __Vehicle Size__
      * Self explanatory
  - __Vehicle Style__
      * Refers to their body style, We are familiar with cars without roofs (Convertible), normal cars (Sedans) and many more
  - __Highway MPG__
      * Refers to Miles Per Gallons, and it is used to show how far the car can travel for every gallon on an highway. (Travelling on highway does not require the car to Stop and Accelerate constantly. Therefore will usually result in Higher MPG compared to City MPG)
  - __City MPG__
      * Similar to Highway MPG, but shows how far the car travels for every gallon in the City.
  - __Popularity__
      * Data was scraped from twitter polls to indicate the popularity of the car.
  - __MSRP__
      * Refers to the manufacturer's suggested retail price
    

Find Link to Sections here <br>
   - [1.Import Libraries](#Importing)
   - [2.Data Extraction](#Data)
   - [3.Checking Data](#Check_Data)
   
  

<hr>

# 1. Import Libraries

> Importing libraries for EDA

<a id = "Importing"></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

<hr>

# 2. Data Extraction 

<a id = "Data"></a>

In [3]:
car_raw = pd.read_csv('Car.csv')

car_raw.sample(10)

Make             Model  Year                  Engine Fuel Type  \
5825    Chevrolet            Impala  2015  flex-fuel (unleaded/natural gas)   
11268      Toyota             Venza  2013                  regular unleaded   
8067        Dodge           RAM 150  1992                  regular unleaded   
1477    Chevrolet       Astro Cargo  2003                  regular unleaded   
4758         Ford              Flex  2015    premium unleaded (recommended)   
4932         Ford         Freestyle  2005                  regular unleaded   
10325        Ford             Tempo  1993                  regular unleaded   
6063   Volkswagen  Jetta SportWagen  2012                  regular unleaded   
2520     Chrysler            Cirrus  1999                  regular unleaded   
5873       Subaru           Impreza  2016                  regular unleaded   

       Engine HP  Engine Cylinders Transmission Type      Driven_Wheels  \
5825         NaN               6.0         AUTOMATIC  front wheel drive   
11268      181.0               4.0         AUTOMATIC  front wheel drive   
8067       230.0               8.0            MANUAL   rear wheel drive   
1477       190.0               6.0         AUTOMATIC    all wheel drive   
4758       365.0               6.0         AUTOMATIC    all wheel drive   
4932       203.0               6.0         AUTOMATIC  front wheel drive   
10325       96.0               4.0            MANUAL  front wheel drive   
6063       170.0               5.0         AUTOMATIC  front wheel drive   
2520       168.0               6.0         AUTOMATIC  front wheel drive   
5873       148.0               4.0            MANUAL    all wheel drive   

       Number of Doors        Market Category Vehicle Size  \
5825               4.0  Flex Fuel,Performance        Large   
11268              4.0              Crossover      Midsize   
8067               2.0                    NaN        Large   
1477               3.0                    NaN        Large   
4758               4.0  Crossover,Performance        Large   
4932               4.0              Crossover        Large   
10325              4.0                    NaN      Compact   
6063               4.0                    NaN      Compact   
2520               4.0                    NaN      Midsize   
5873               4.0                    NaN      Compact   

             Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
5825                 Sedan           25        17        1385  40660  
11268                Wagon           26        20        2031  31510  
8067   Extended Cab Pickup           16        11        1851   2000  
1477         Cargo Minivan           18        14        1385  24450  
4758                 Wagon           23        16        5657  42400  
4932                 Wagon           25        18        5657  26440  
10325                Sedan           30        21        5657   2000  
6063                 Wagon           31        24         873  25960  
2520                 Sedan           25        17        1013   2202  
5873                 Sedan           34        25         640  18295

<hr>

# 3. Checking Data

<a id = "Check_Data"></a>

In [4]:
car_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

It appears that we have 16 features and 11914 entries. However the dataset seems to be incomplete with 


In [16]:
null_col = car_raw.columns[car_raw.isnull().any()]
car_raw[null_col].isnull().sum()

Engine Fuel Type       3
Engine HP             69
Engine Cylinders      30
Number of Doors        6
Market Category     3742
dtype: int64

>Checking Engine Fuel Type. Due to imcomplete information we will drop these columns

In [17]:
null = car_raw[car_raw.isna().any(axis=1)]
car_raw[car_raw['Engine Fuel Type'].isna()]

Make   Model  Year Engine Fuel Type  Engine HP  Engine Cylinders  \
11321  Suzuki  Verona  2004              NaN      155.0               6.0   
11322  Suzuki  Verona  2004              NaN      155.0               6.0   
11323  Suzuki  Verona  2004              NaN      155.0               6.0   

      Transmission Type      Driven_Wheels  Number of Doors Market Category  \
11321         AUTOMATIC  front wheel drive              4.0             NaN   
11322         AUTOMATIC  front wheel drive              4.0             NaN   
11323         AUTOMATIC  front wheel drive              4.0             NaN   

      Vehicle Size Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
11321      Midsize         Sedan           25        17         481  17199  
11322      Midsize         Sedan           25        17         481  20199  
11323      Midsize         Sedan           25        17         481  18499

In [29]:
car_clean = car_raw.drop([11321, 11322, 11323])
car_clean[null_col].isnull().sum()

Engine Fuel Type       0
Engine HP             69
Engine Cylinders      30
Number of Doors        6
Market Category     3739
dtype: int64

>Checking Engine HP we can see that most electric cars do not have their HP allocated.

In [18]:
car_raw[car_raw['Engine HP'].isna()]

Make        Model  Year                Engine Fuel Type  Engine HP  \
539      FIAT         500e  2015                        electric        NaN   
540      FIAT         500e  2016                        electric        NaN   
541      FIAT         500e  2017                        electric        NaN   
2905  Lincoln  Continental  2017  premium unleaded (recommended)        NaN   
2906  Lincoln  Continental  2017  premium unleaded (recommended)        NaN   
...       ...          ...   ...                             ...        ...   
9850      Kia      Soul EV  2015                        electric        NaN   
9851      Kia      Soul EV  2015                        electric        NaN   
9852      Kia      Soul EV  2016                        electric        NaN   
9853      Kia      Soul EV  2016                        electric        NaN   
9854      Kia      Soul EV  2016                        electric        NaN   

      Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
539                0.0      DIRECT_DRIVE  front wheel drive              2.0   
540                0.0      DIRECT_DRIVE  front wheel drive              2.0   
541                0.0      DIRECT_DRIVE  front wheel drive              2.0   
2905               6.0         AUTOMATIC    all wheel drive              4.0   
2906               6.0         AUTOMATIC  front wheel drive              4.0   
...                ...               ...                ...              ...   
9850               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9851               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9852               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9853               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9854               0.0      DIRECT_DRIVE  front wheel drive              4.0   

     Market Category Vehicle Size  Vehicle Style  highway MPG  city mpg  \
539        Hatchback      Compact  2dr Hatchback          108       122   
540        Hatchback      Compact  2dr Hatchback          103       121   
541        Hatchback      Compact  2dr Hatchback          103       121   
2905          Luxury        Large          Sedan           25        17   
2906          Luxury        Large          Sedan           27        18   
...              ...          ...            ...          ...       ...   
9850             NaN      Compact          Wagon           92       120   
9851             NaN      Compact          Wagon           92       120   
9852             NaN      Compact          Wagon           92       120   
9853             NaN      Compact          Wagon           92       120   
9854             NaN      Compact          Wagon           92       120   

      Popularity   MSRP  
539          819  31800  
540          819  31800  
541          819  31800  
2905          61  55915  
2906          61  62915  
...          ...    ...  
9850        1720  35700  
9851        1720  33700  
9852        1720  33950  
9853        1720  31950  
9854        1720  35950  

[69 rows x 16 columns]

>Checking Engine Cylinders we can see that all electric cars are missing values. This is due to electric vehicles not having engines.

>Another interesting vehicle model that suffers from this is the RX7 and RX8. This is due to the vehicles having rotary engines that do not have cylinders.

>Therefore we should create a new category R.

In [21]:
car_raw[car_raw['Engine Cylinders'].isna()]

Make    Model  Year             Engine Fuel Type  Engine HP  \
1983   Chevrolet  Bolt EV  2017                     electric      200.0   
1984   Chevrolet  Bolt EV  2017                     electric      200.0   
3716  Volkswagen   e-Golf  2015                     electric      115.0   
3717  Volkswagen   e-Golf  2015                     electric      115.0   
3718  Volkswagen   e-Golf  2016                     electric      115.0   
3719  Volkswagen   e-Golf  2016                     electric      115.0   
5778  Mitsubishi   i-MiEV  2014                     electric        NaN   
5779  Mitsubishi   i-MiEV  2016                     electric       66.0   
5780  Mitsubishi   i-MiEV  2017                     electric       66.0   
8373      Toyota  RAV4 EV  2012                     electric      154.0   
8695       Mazda     RX-7  1993             regular unleaded      255.0   
8696       Mazda     RX-7  1994             regular unleaded      255.0   
8697       Mazda     RX-7  1995             regular unleaded      255.0   
8698       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8699       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8700       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8701       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8702       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8703       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8704       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8705       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8706       Mazda     RX-8  2010  premium unleaded (required)      212.0   
8707       Mazda     RX-8  2010  premium unleaded (required)      212.0   
8708       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8709       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8710       Mazda     RX-8  2011  premium unleaded (required)      212.0   
8711       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8712       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8713       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8714       Mazda     RX-8  2011  premium unleaded (required)      212.0   

      Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
1983               NaN      DIRECT_DRIVE  front wheel drive              4.0   
1984               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3716               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3717               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3718               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3719               NaN      DIRECT_DRIVE  front wheel drive              4.0   
5778               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5779               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5780               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
8373               NaN      DIRECT_DRIVE  front wheel drive              4.0   
8695               NaN            MANUAL   rear wheel drive              2.0   
8696               NaN            MANUAL   rear wheel drive              2.0   
8697               NaN            MANUAL   rear wheel drive              2.0   
8698               NaN            MANUAL   rear wheel drive              4.0   
8699               NaN         AUTOMATIC   rear wheel drive              4.0   
8700               NaN            MANUAL   rear wheel drive              4.0   
8701               NaN            MANUAL   rear wheel drive              4.0   
8702               NaN            MANUAL   rear wheel drive              4.0   
8703               NaN         AUTOMATIC   rear wheel drive              4.0   
8704               NaN         AUTOMATIC   rear wheel drive              4.0   
8705  

In [37]:
if 'electric' in car_clean["Engine Fuel Type"]:
    car_clean = car_clean["Engine Cylinders"].fillna("0", inplace = True)
    
car_clean[null_col].isnull().sum()

Engine Fuel Type       0
Engine HP             69
Engine Cylinders      30
Number of Doors        6
Market Category     3739
dtype: int64

In [22]:
car_raw[car_raw['Number of Doors'].isna()]

Make    Model  Year             Engine Fuel Type  Engine HP  \
4666  Ferrari       FF  2013  premium unleaded (required)      651.0   
6930    Tesla  Model S  2016                     electric        NaN   
6931    Tesla  Model S  2016                     electric        NaN   
6932    Tesla  Model S  2016                     electric        NaN   
6933    Tesla  Model S  2016                     electric        NaN   
6934    Tesla  Model S  2016                     electric        NaN   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
4666              12.0  AUTOMATED_MANUAL   all wheel drive              NaN   
6930               0.0      DIRECT_DRIVE   all wheel drive              NaN   
6931               0.0      DIRECT_DRIVE   all wheel drive              NaN   
6932               0.0      DIRECT_DRIVE   all wheel drive              NaN   
6933               0.0      DIRECT_DRIVE  rear wheel drive              NaN   
6934               0.0      DIRECT_DRIVE   all wheel drive              NaN   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
4666  Exotic,High-Performance        Large         Coupe           16   
6930       Exotic,Performance        Large         Sedan          105   
6931       Exotic,Performance        Large         Sedan          101   
6932  Exotic,High-Performance        Large         Sedan          105   
6933       Exotic,Performance        Large         Sedan          100   
6934       Exotic,Performance        Large         Sedan          107   

      city mpg  Popularity    MSRP  
4666        11        2774  295000  
6930       102        1391   79500  
6931        98        1391   66000  
6932        92        1391  134500  
6933        97        1391   74500  
6934       101        1391   71000

In [23]:
car_raw[car_raw['Market Category'].isna()]

Make     Model  Year                Engine Fuel Type  Engine HP  \
87     Nissan     200SX  1996                regular unleaded      115.0   
88     Nissan     200SX  1996                regular unleaded      115.0   
91     Nissan     200SX  1997                regular unleaded      115.0   
92     Nissan     200SX  1997                regular unleaded      115.0   
93     Nissan     200SX  1998                regular unleaded      115.0   
...       ...       ...   ...                             ...        ...   
11794  Subaru        XT  1991                regular unleaded      145.0   
11809  Toyota  Yaris iA  2017                regular unleaded      106.0   
11810  Toyota  Yaris iA  2017                regular unleaded      106.0   
11867     GMC     Yukon  2015  premium unleaded (recommended)      420.0   
11868     GMC     Yukon  2015  premium unleaded (recommended)      420.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
87                  4.0            MANUAL  front wheel drive              2.0   
88                  4.0            MANUAL  front wheel drive              2.0   
91                  4.0            MANUAL  front wheel drive              2.0   
92                  4.0            MANUAL  front wheel drive              2.0   
93                  4.0            MANUAL  front wheel drive              2.0   
...                 ...               ...                ...              ...   
11794               6.0            MANUAL    all wheel drive              2.0   
11809               4.0            MANUAL  front wheel drive              4.0   
11810               4.0         AUTOMATIC  front wheel drive              4.0   
11867               8.0         AUTOMATIC   rear wheel drive              4.0   
11868               8.0         AUTOMATIC   four wheel drive              4.0   

      Market Category Vehicle Size Vehicle Style  highway MPG  city mpg  \
87                NaN      Compact         Coupe           36        26   
88                NaN      Compact         Coupe           36        26   
91                NaN      Compact         Coupe           35        25   
92                NaN      Compact         Coupe           35        25   
93                NaN      Compact         Coupe           35        25   
...               ...          ...           ...          ...       ...   
11794             NaN      Compact         Coupe           23        16   
11809             NaN      Compact         Sedan           39        30   
11810             NaN      Compact         Sedan           40        32   
11867             NaN        Large       4dr SUV           21        15   
11868             NaN        Large       4dr SUV           21        14   

       Popularity   MSRP  
87           2009   2000  
88           2009   2000  
91           2009   2000  
92           2009   2000  
93           2009   2000  
...           ...    ...  
11794         640   2000  
11809        2031  15950  
11810        2031  17050  
11867         549  64520  
11868         549  67520  

[3742 rows x 16 columns]